Internuncio/open-guilds
# Amazon integration sc

So I removed most of what you originally put in. I didn't like how messy the separate contracts and received data was. In short:

- Datums have a priority determined by the completion percentage of a batch and the total time the datum has been waiting to be processed
- Layers have different services (right now either amazon mechanical turk or open guilds)
- Layers manage a priority queue of data and a pool of already queued contracts
- When the contracts dip below a certain integer it will post in whichever queue is available
- Contracts get created and then users update themselves to be assigned to it, same thing with amazon mechanical turk workers
- AMT hit_id and hit_url are stored in external_url and external_id, which will be extensible to other services in the future, but may eventually deserve its own object.

I went down this route for some key reasons:
1. Queueing theory is going to be central in us being able to control cost vs time for data completion
2. Therefore we need a way to get Little's equations to get best estimates of completion time and approximations of load handling for guilds.
3. In the future I want to allow layers to be formed of other guilds, so as long as they follow the same queueing protocol it should be easy to switch over.


To test the feature:
1. Sign in as a user
2. Create a guild
3. Fill in your credentials for AWS under My Account
4. Create a layer and check Amazon Mechanical Turk under services
5. Add a batch with a datum
6. Run rake jobs:work in a terminal instance to post the jobs to AMT
7. Go to the batch and show the datum
8. Click on "Link" and sign in to AMT workers sandbox to see your task successfully posted to AMT.
